### PR TITLE
Requested service resources adaptation

### PR DESCRIPTION
**Pull request by mykola-kobets-epam**

### CODE REVIEW

File: launcher/launcher.go [hunk 1]
```diff
@@ -687,46 +687,70 @@ func (launcher *Launcher) getLayersForService(digests []string) ([]imagemanager.
...
 	}
 
-	return defaultResourceRation / 100
+	return 0
+}
+
+func clampResource(value uint64, quota *uint64) uint64 {
+	if quota != nil && value > *quota {
+		return *quota
+	} else {
```
Comment (mlohvynenko) [inline launcher/launcher.go:751]: Let's simple return value, without else branch

File: launcher/launcher.go [hunk 2]
```diff
@@ -687,46 +687,70 @@ func (launcher *Launcher) getLayersForService(digests []string) ([]imagemanager.
...
+		stateSize = clampResource(*serviceConfig.RequestedResources.Storage, stateQuota)
+		storageSize = clampResource(*serviceConfig.RequestedResources.Storage, storageQuota)
+	} else {
+		stateSize = getReqStorageFromNodeConf(stateQuota, nodeRatios)
+		storageSize = getReqStorageFromNodeConf(storageQuota, nodeRatios)
 	}
 
-	if nodeRatios != nil && nodeRatios.Storage != nil {
-		return float64(*nodeRatios.Storage) / 100
+	return
```
Comment (mlohvynenko) [inline launcher/launcher.go:703]: Is this return required?

Comment (mykola-kobets-epam) [reply launcher/launcher.go:703]: yes, "missing return " error have otherwise

File: launcher/launcher.go (hunk 1 above)
Comment (mykola-kobets-epam) [reply launcher/launcher.go]: fixed

**Review by mlohvynenko — APPROVED** [review]

File: launcher/launcher.go [hunk 3]
```diff
@@ -687,46 +687,70 @@ func (launcher *Launcher) getLayersForService(digests []string) ([]imagemanager.
...
-) float64 {
-	if serviceRatios != nil && serviceRatios.Storage != nil {
-		return float64(*serviceRatios.Storage) / 100
+func getReqDiskSize(serviceConfig aostypes.ServiceConfig, nodeRatios *aostypes.ResourceRatiosInfo,
+) (stateSize, storageSize uint64) {
+	stateQuota := serviceConfig.Quotas.StateLimit
+	storageQuota := serviceConfig.Quotas.StorageLimit
+
+	if serviceConfig.RequestedResources != nil && serviceConfig.RequestedResources.Storage != nil {
+		stateSize = clampResource(*serviceConfig.RequestedResources.Storage, stateQuota)
```
Comment (al1img) [inline launcher/launcher.go:696]: Clamping should be done for all the cases.

File: launcher/nodehandler.go [hunk 4]
```diff
@@ -311,10 +311,12 @@ func (node *nodeHandler) getRequestedCPU(
 	instanceIdent aostypes.InstanceIdent, serviceConfig aostypes.ServiceConfig,
 ) uint64 {
 	requestedCPU := uint64(0)
+	cpuQuota := serviceConfig.Quotas.CPUDMIPSLimit
```
Comment (al1img) [inline launcher/nodehandler.go:314]: ditto for the intermediate variable and clamping

File: launcher/nodehandler.go [hunk 5]
```diff
@@ -336,10 +338,12 @@ func (node *nodeHandler) getRequestedRAM(
 	instanceIdent aostypes.InstanceIdent, serviceConfig aostypes.ServiceConfig,
 ) uint64 {
 	requestedRAM := uint64(0)
+	ramQuota := serviceConfig.Quotas.RAMLimit
```
Comment (al1img) [inline launcher/nodehandler.go:341]: ditto

File: launcher/launcher.go [hunk 6]
```diff
@@ -687,46 +687,70 @@ func (launcher *Launcher) getLayersForService(digests []string) ([]imagemanager.
...
-	serviceRatios *aostypes.ResourceRatiosInfo, nodeRatios *aostypes.ResourceRatiosInfo,
-) float64 {
-	if serviceRatios != nil && serviceRatios.Storage != nil {
-		return float64(*serviceRatios.Storage) / 100
+func getReqDiskSize(serviceConfig aostypes.ServiceConfig, nodeRatios *aostypes.ResourceRatiosInfo,
+) (stateSize, storageSize uint64) {
+	stateQuota := serviceConfig.Quotas.StateLimit
+	storageQuota := serviceConfig.Quotas.StorageLimit
+
+	if serviceConfig.RequestedResources != nil && serviceConfig.RequestedResources.Storage != nil {
```
Comment (al1img) [inline launcher/launcher.go:695]: We should have different values for state and storage. Let's discuss with cloud.

Comment (mykola-kobets-epam) [reply launcher/launcher.go:701]: fixed

File: launcher/launcher.go (hunk 2 above)
Comment (mykola-kobets-epam) [reply launcher/launcher.go]: fixed

File: launcher/launcher.go [hunk 7]
```diff
@@ -687,46 +687,70 @@ func (launcher *Launcher) getLayersForService(digests []string) ([]imagemanager.
...
 }
 
-func getStorageRequestRatio(
-	serviceRatios *aostypes.ResourceRatiosInfo, nodeRatios *aostypes.ResourceRatiosInfo,
-) float64 {
-	if serviceRatios != nil && serviceRatios.Storage != nil {
-		return float64(*serviceRatios.Storage) / 100
+func getReqDiskSize(serviceConfig aostypes.ServiceConfig, nodeRatios *aostypes.ResourceRatiosInfo,
+) (stateSize, storageSize uint64) {
+	stateQuota := serviceConfig.Quotas.StateLimit
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply launcher/launcher.go:692]: to make a line shorter in a place where it is used so it didnt move to a next line

[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply launcher/launcher.go:692]: to make a line shorter in a place where it is used so it didnt move to a next line

File: launcher/launcher.go (hunk 3 above)
Comment (mykola-kobets-epam) [reply launcher/launcher.go]: It's not required to clamp values from NodeConfig as they are a ratios of quota,

File: launcher/nodehandler.go (hunk 4 above)
Comment (al1img) [reply launcher/nodehandler.go:314]: Somehow my comments about clamping for all cases disappeared: clamping should be done for all cases.

File: launcher/launcher.go [hunk 8]
```diff
@@ -687,46 +687,88 @@ func (launcher *Launcher) getLayersForService(digests []string) ([]imagemanager.
...
+	storageQuota := serviceConfig.Quotas.StorageLimit
+
+	if serviceConfig.RequestedResources != nil && serviceConfig.RequestedResources.State != nil {
+		stateSize = clampResource(*serviceConfig.RequestedResources.State, stateQuota)
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply launcher/launcher.go:696]: in node config resource ratios specified in the percents of the quota. that is by default should be less than resource quota

File: launcher/nodehandler.go (hunk 4 above)
Comment (mykola-kobets-epam) [reply launcher/nodehandler.go:314]: fixed

**Pull request merged**